kernel/src/grant.rs

```diff
@@ -170,6 +170,46 @@ impl<'a, T: 'a + ?Sized> DerefMut for GrantData<'a, T> {
     }
 }
 
+/// Errors which can occur when scheduling a process Upcall
+///
+/// Scheduling a null-Upcall (which will not be delivered to a
+/// process) is deliberately not an error, given that a null-Upcall is
+/// a well-defined Upcall to be set by a process. It behaves
+/// essentially the same as if the process would set a proper Upcall,
+/// and would ignore all invocations, with the benefit that no task is
+/// inserted in the process' task queue.
+#[derive(Copy, Clone, Debug)]
+pub enum UpcallError {
+    /// The passed `subscribe_num` exceeds the number of Upcalls
+    /// available for this process.
+    ///
+    /// For a [`Grant`] with `n` `NUM_UPCALLS`, this error is returned
+    /// when `GrantUpcallTable::schedule_upcall` is invoked with
+    /// `subscribe_num >= n`.
+    ///
+    /// No Upcall has been scheduled, the call to
+    /// `GrantUpcallTable::schedule_upcall` had no observable effects.
+    ///
+    InvalidSubscribeNum,
+    /// The process' task queue is full.
+    ///
+    /// This error can occur when too many tasks (for example,
+    /// Upcalls) have been scheduled for a process, without that
+    /// process yielding or having a chance to resume execution.
+    ///
+    /// No Upcall has been scheduled, the call to
+    /// `GrantUpcallTable::schedule_upcall` had no observable effects.
+    QueueFull,
+    /// A kernel-internal invariant has been violated.
+    ///
+    /// This error should never happen. It can be returned if the
+    /// process is inactive (which should be caught by
+    /// [`Grant::enter`]) or `process.tasks` was taken.
+    ///
+    /// These cases cannot be reasonably handled.
+    KernelError,
```

Comment:
Is this error worth including? I am imaging me as a handler of this code -- what do I do? The options are
 1. Just ignore the error and hope (same as if an error had not been returned)
 2. `panic!`, the world is broken

I think that making capsule authors choose between these will likely just result in a lot of arbitrary coin flips between these two options.

Reply:
I suppose the issue is that `panic!`ing in the kernel is not an option (even if these error cases were to occur, the system might not be in an entirely unrecoverable state), whereas returning `Ok(())` as if the callback was scheduled is obviously misleading.

Reply:
It's convenient to include the error, because rust forces us to handle cases that can "never" happen. I think it is better to have the error than do something like hardcode panics into the core kernel binary which will never be triggered, but still take up space.

From a capsule author's point of view, I think it's pretty clear that it's not an error they should handle, but could report upstream to us if they ever saw it.

However, we do not want capsule authors to include a debug or panic, because that just re-creates the size issue (and probably worse). But that is also true of handling this return value at all, unless a capsule can do something meaningful it should just ignore the return value.